Mathlib/Algebra/Lie/Killing.lean

```diff
@@ -0,0 +1,94 @@
+/-
+Copyright (c) 2023 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash
+-/
+import Mathlib.Algebra.Lie.Nilpotent
+import Mathlib.LinearAlgebra.Trace
+
+/-!
+# The trace and Killing forms of a Lie algebra.
+
+Let `L` be a Lie algebra with coefficients in a commutative ring `R`. Suppose `M` is a finite, free
+`R`-module and we have a representation `φ : L → End M`. This data induces a natural bilinear form
+`B` on `L`, called the trace form associated to `M`; it is defined as `B(x, y) = Tr (φ x) (φ y)`.
+
+In the special case that `M` is `L` itself and `φ` is the adjoint representation, the trace form
+is known as the Killing form.
+
+We define the trace / Killing form in this file and prove some basic properties.
+
+## Main definitions
+
+ * `LieModule.traceForm`
+ * `killingForm`
+
+## TODO
+
+ * Show that `LieModule.traceForm R L M` vanishes when `M` is nilpotent (using
+   `isNilpotent_toEndomorphism_of_isNilpotent₂`).
+ * Prove Cartan's criterion for semisimplicity.
+-/
+
+variable (R L : Type _) [CommRing R] [LieRing L] [LieAlgebra R L]
+
+namespace LieModule
+
+open LinearMap (trace)
+
+variable (M : Type _) [AddCommGroup M] [Module R M] [LieRingModule L M] [LieModule R L M]
+variable [Module.Free R M] [Module.Finite R M]
+
+local notation "φ" => toEndomorphism R L M
+
+/-- A finite, free representation of a Lie algebra `L` induces a bilinear form on `L` called
+the trace Form. See also `killingForm`. -/
+noncomputable def traceForm : L →ₗ[R] L →ₗ[R] R :=
+{ toFun := fun x ↦
+    { toFun := fun y ↦ trace R _ (φ x ∘ₗ φ y)
+      map_add' := fun y z ↦ by simp [LinearMap.comp_add]
+      map_smul' := fun t y ↦ by simp [LinearMap.comp_smul] }
+  map_add' := fun x y ↦ by ext z; simp [LinearMap.add_comp]
+  map_smul' := fun t x ↦ by ext y; simp [LinearMap.smul_comp] }
```

Comment:
`LinearMap.mk\2` is probably sensible to use here. Also, don't forget we now indent definitions by two spaces!

Reply:
```suggestion
  ((LinearMap.mul _ _).compl₁₂ (φ).toLinearMap (φ).toLinearMap).compr₂ (trace R M)
```

Reply:
Impressive!